### PR TITLE
Update to release-tools for the 23.10 release

### PR DESCRIPTION
**Pull request by ajkavanagh**
These updates are:

 * Updates to global/* files to support pip-compile; this changes the
   defaults to *.in files and updates tox.ini to work with these files.
   Note that the k8s version have been left unaltered.
 * Added 'lock-with-pip-compile' that migrates existing *.txt
   requirements files to *.in in the repos and then performs the
   pip-compile by merging *.in files; during the next cycle we should
   rationalise the requirements files so they don't need to be merged.
 * Moved fetch-charms.py out of the tox as running it via tox does very
   strange things to the git command (and shell commands in general).
   Added a fetch-charms script that checkes for a py3 venv, makes it if
   it doesn't exist (using tox) and then runs fetch-charms.py using the
   python from that venv.
 * Other updates to helper scripts that were useful for the stablisation
   reviews.

### CODE REVIEW

File: global/classic-zaza/tox.ini [hunk 1]
```diff
@@ -47,15 +47,17 @@ commands =
...
+
 [testenv:py310]
 basepython = python3.10
-deps = -r{toxinidir}/requirements.txt
-       -r{toxinidir}/test-requirements.txt
+deps = -r{toxinidir}/merged-requirements-py310.txt
 
 [testenv:py3]
 basepython = python3
-deps = -r{toxinidir}/requirements.txt
```
Comment (coreycb) [inline global/classic-zaza/tox.ini:57]: Out of curiosity why do the default py3 targets use py38 instead of py310?

**Review by coreycb — COMMENTED** [review]
This looks good to me. I did have one question out of curiosity inline.

File: global/classic-zaza/tox.ini (hunk 1 above)
Comment (ajkavanagh) [reply global/classic-zaza/tox.ini:57]: It's a good question, and one I pondered for a bit.  Essentially, I went for the lowest common denominator; the py3 target isn't used in the CI system(s) (they use the py38 and py310 targets directly), so it is essentially a fallback target.  I did think about removing it, but considered that even more painful!

**Review by freyes — APPROVED** [review]
LGTM. It's a large set of changes, lot of work, thanks Alex.

@coreycb , are you good with the Alex's response to "why use py38 deps"?

**Comment by coreycb** [comment]
@freyes Yes I'm ok with it. It seems it's needed for the greater good, and developers can either use https://launchpad.net/~deadsnakes or update tox.ini when testing locally.

**Pull request merged**